Snapping update

mtschoen-unity opened the pull request:
Timoni noticed a couple of unexpected behaviors when testing the snapping feature.

Closes #192, Closes #195, Closes #193 

The issue with the sphere manipulator was actually intended behavior, but we decided that it was best to switch "Limit snapping radius" to be enabled by default, so that objects won't appear to "fly away" unexpectedly.

The linear mainipulator handles simply didn't have a snapping solution, so that was added. Snap to ground was implemented, but too easy to "punch through."

In testing these changes, we discovered a need to add a new "punch through" requirement to the design spec. On top of the rule that it should require more distance break snapping than to snap in the first place, it should also require a much greater break distance if you are moving against the surface normal. This "punch through" break distance is now set to 5x the normal break distance.

In the process, I ended up cleaning up the snapping code a little bit. Because this "punch through" is such a different approach, it is possible we have created new edge cases, so thorough testing is recommended.

amirebrahimi commented on Scripts/Interfaces/Entity/IManipulator.cs:
```diff
@@ -13,7 +21,7 @@ public interface IManipulator
 		/// Delegate that processes the translation, using the vector3 passed in
 		/// Caller also provides the ray origin that is doing the action, and whether or not this translation is axis-constrained
```
Probably want to update the comments to reflect the change from bool to ConstrainedAxis. Something to the effect of "and constrained axes"

amirebrahimi commented on Manipulators/StandardManipulator.cs:
```diff
@@ -67,7 +67,16 @@ void Update()
 		{
-			translate(eventData.deltaPosition, eventData.rayOrigin, !(handle is SphereHandle));
+			ConstrainedAxis constraints = 0;
+			var linear = handle as LinearHandle;
```
Create an internal interface called IAxisConstraints that exposes the property, so you can simply check for that to get the constraints off instead of checking for specific classes.

amirebrahimi commented on Scripts/Interfaces/Entity/IManipulator.cs:
```diff
@@ -4,6 +4,14 @@
 namespace UnityEditor.Experimental.EditorVR
 {
+	[Flags]
+	public enum ConstrainedAxis
```
Move to a separate file

mtschoen-unity commented:
@amirebrahimi I've addressed your comments and added one final touch: scaling the ground plane based on viewer scale. I was noticing that you can see the edges when you scale yourself very big. The problem with just using a larger scale all the time is that it flickers at regular scales with a scale of 1000000

amirebrahimi commented on Scripts/Core/Interfaces/IAxisConstraints.cs:
```diff
@@ -0,0 +1,9 @@
+﻿#if UNITY_EDITOR
+namespace UnityEditor.Experimental.EditorVR
+{
+	interface IConstrainedAxis
```
Please rename to IAxisConstraints

amirebrahimi commented on Scripts/Data/ConstrainedAxisFlags.cs:
```diff
@@ -0,0 +1,17 @@
+	/// Which axes are constrained
+	/// </summary>
+	[Flags]
+	public enum ConstrainedAxis
```
Either rename this enum to ConstrainedAxisFlags or the file to ConstrainedAxis.cs.

AndrewTHEManeri commented on Scripts/Modules/SnappingModule/SnappingModule.cs:
```diff
@@ -511,65 +526,201 @@ public bool DirectSnap(Transform rayOrigin, Transform transform, ref Vector3 pos
+		static bool TryBreakSnap(ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, Quaternion targetRotation, SnappingState state, float breakDistance)
 		{
-			if (state.surfaceSnapping)
+			if (state.snapping)
```
Would really fast motion that stayed within snapping range possible trigger a snap break?

AndrewTHEManeri commented on Tools/TransformTool/TransformTool.cs:
```diff
@@ -537,12 +539,22 @@ public void DropHeldObjects(Node node)
+			{
+				case ConstrainedAxis.X | ConstrainedAxis.Y:
+				case ConstrainedAxis.Y | ConstrainedAxis.Z:
+				case ConstrainedAxis.X | ConstrainedAxis.Z:
```
Should the axis constraints be adjusting the delta here to knock out the constrained/non-constrained axes?

mtschoen-unity replied:
No. This happens in the handle itself. The reason we are doing this is that we don't actually have a solution for snapping the planar handles yet. My original approach was just to treat them as a snap in one axis, then a snap in the other, but that felt unpredictable. When I discussed this with Timoni, we agreed that since the planar handles already "snap" to a plane, it is not as important that they snap to surfaces.
